rosbag2_py/test/test_transport.py

```diff
@@ -40,3 +46,50 @@ def test_options_qos_conversion():
     record_options = rosbag2_py.RecordOptions()
     record_options.topic_qos_profile_overrides = simple_overrides
     assert record_options.topic_qos_profile_overrides == simple_overrides
+
+
+def test_record_cancel(tmp_path):
+    bag_path = str(tmp_path / 'test_record_cancel')
+    storage_options, converter_options = get_rosbag_options(bag_path)
+
+    recorder = rosbag2_py.Recorder()
+
+    record_options = rosbag2_py.RecordOptions()
+    record_options.all = True
+    record_options.is_discovery_disabled = False
+    record_options.rmw_serialization_format = ""
+    record_options.topic_polling_interval = datetime.timedelta(milliseconds=100)
+
+    record_options.compression_mode = 'file'
+    record_options.compression_queue_size = 1
+    record_options.compression_threads = 0
+
+    rclpy.init()
+    record_thread = threading.Thread(
+        target=recorder.record,
+        args=(storage_options, record_options),
+        daemon=True)
+    record_thread.start()
+
+    node = rclpy.create_node('test_record_cancel')
+    executor = rclpy.executors.SingleThreadedExecutor()
+    executor.add_node(node)
+    pub = node.create_publisher(String, 'chatter', 10)
+
+    i = 0
+    import time
+    while rclpy.ok() and i < 10:
+        msg = String()
+        msg.data = 'Hello World: {0}'.format(i)
+        i += 1
+        pub.publish(msg)
+        time.sleep(0.1)
+
+    recorder.cancel()
+    msg = String()
+    msg.data = 'Spining one last time'
+    pub.publish(msg)
```

Comment:
why is it necessary to "spin one more time"? And, instead of sleeping for an arbitrary time, could you instead wait for the expected files to exist, and fail on some timeout? That should be more reliable than hoping that "0.1 seconds is probably long enough"

Reply:
I noticed if you are not publishing anything when you call the ```recorder.cancel()``` the ```spin_until_future_complete``` stays blocked until there is another spin. That's why I added one last message to be sent so the thread will finish properly creating the metafile.yaml . Removing the sleep of line 86 might also prevent this behavior because when the cancel command is activated there might be some messages in the buffer. However, I think is wise to leave the spining one last time message. 
This behavior can be reproduced if you are calling the recorder and cancel without any publishers.

Reply:
This _should_ work without new spinning/events coming in. The `spin_until_future_complete` loop is supposed to wake "immediately" when the future is marked complete, in both rclpy and rclcpp.

I'm going to run a quick experiment locally. I wonder if it has something to do with the GIL or something.